### ISSUE TITLE
epoch and iteration, which one is more importand

### ISSUE THREAD

**Issue by jimwangzx** [AlexeyAB/darknet#2513]
As suggested, one class needs 2000 iterations. But the number of images may vary. Why we emphasize iterations rather than epoch?

**Comment by AlexeyAB** [comment]
@jimwangzx 


* If you have images < 2000 per class, then the iterations number is more important
* If you have images > 2000 per class, then the epoch number is more important

----

https://github.com/AlexeyAB/darknet#how-to-improve-object-detection

> you should preferably have 2000 different images for each class or more, and you should train **2000*classes iterations or more**

https://github.com/AlexeyAB/darknet#when-should-i-stop-training

> Usually sufficient 2000 iterations for each class(object), but **not less than 4000 iterations in total**. But for a more precise definition when you should stop training, use the following manual:


In general, you should train from `4000` iterations to `2000*classes` iterations in total **or more** (while your mAP increases).

**There are just empirical conclusions:**

* if you even train for 10 images and 1 class, then the good detection you will achive after 4000 iterations

* if you train for 2 000 images per class for 20 classes (40 000 images in total), then you should train until your mAP increases,  usually 40 000 iterations are enough to train (if you can't / don't want to check mAP)

* if you train for 10 000 images per class for 20 classes (200 000 images in total), then you should train until your mAP increases, it can be 200 000 iterations or 2 000 000 iterations


----

May be we can use such very rough calculations:
Since we use `learning_rate=0.001` = 1/1000, then we should train at least ~1000 iterations for each final activation.
But considering that you can have images for one class 2x time more than for another class, we just will train 2000 iterations instead of 1000 iterations.

Since there is data augmentation, then the different number of images will be much more than you have.

**Comment by jimwangzx** [comment]
@AlexeyAB 
Thanks
There are just empirical conclusions:

> if you even train for 10 images and 1 class, then the good detection you will achive after 4000 iterations
> 
> if you train for 2 000 images per class for 20 classes (40 000 images in total), then you should train until your mAP increases, usually 40 000 iterations are enough to train (if you can't / don't want to check mAP)
> 
> if you train for 10 000 images per class for 20 classes (200 000 images in total), then you should train until your mAP increases, it can be 200 000 iterations or 2 000 000 iterations

Does this mean the more images, the longer to train to get same result? If so, why we need more images?

**Comment by AlexeyAB** [comment]
@jimwangzx 

> Does this mean the more images, the longer to train to get same result? 

No.

* The same images with more iterations = to get the same results
* The more images with the same iterations = to get the same results
* The more more images + more iterations = to get better results

**Comment by jimwangzx** [comment]
Thanks

**Comment by ydixon** [comment]
@AlexeyAB 

> Since we use learning_rate=0.001 = 1/1000, then we should train at least ~1000 iterations for each final activation.

Is it because we are only updating 1/1000 of the gradients calculated, so we should do around 1000 iterations? 

What does **"each final activation"** means here? Or does it just mean 1 class roughly?

I'm sorry for these trivial questions, I'm kinda lacking in the math department. Thanks!

**Comment by AlexeyAB** [comment]
@ydixon 

> Is it because we are only updating 1/1000 of the gradients calculated, so we should do around 1000 iterations?

Yes.

> What does "each final activation" means here? Or does it just mean 1 class roughly?

It means the output of the last convolutional-layer.

**Comment by ydixon** [comment]
@AlexeyAB  Thank you
